File: exts/jphp-gui-ext/src/main/resources/JPHP-INF/sdk/php/gui/printing/UXPrinter.php
<?php
namespace php\gui\printing;

use php\gui\UXNode;
use php\gui\UXWindow;

/**
 * @packages gui, javafx
 */
class UXPrinter
{
    /**
     * @readonly
     * @var string
     */
    public $name;

    /**
     * @readonly
     * @var array
     */
    public $attributes = [];

    private function __construct()
    {
    }

    /**
     * @return UXPrinterJob
     */
    public function createPrintJob()
    {
    }

    /**
     * @param UXNode $node
     * @return bool
     */
    public function print(UXNode $node)
    {
    }

    /**
     * @param UXWindow $ownerWindow
     * @param UXNode $node
     * @return bool
     */
    public function printWithDialog(UXWindow $ownerWindow, UXNode $node)
    {
    }

    /**
     * @return UXPrinter
     */
    public static function getDefault()
    {
    }

    /**
     * @return UXPrinter[]
     */
    public static function getAll()
    {
    }
}

File: exts/jphp-gui-ext/src/main/java/org/develnext/jphp/ext/javafx/classes/printing/UXPrinter.java
package org.develnext.jphp.ext.javafx.classes.printing;

import javafx.print.Printer;
import javafx.print.PrinterAttributes;
import javafx.print.PrinterJob;
import javafx.scene.Node;
import javafx.stage.Window;
import org.develnext.jphp.ext.javafx.JavaFXExtension;
import php.runtime.annotation.Reflection;
import php.runtime.annotation.Reflection.Getter;
import php.runtime.annotation.Reflection.Property;
import php.runtime.annotation.Reflection.Signature;
import php.runtime.env.Environment;
import php.runtime.env.TraceInfo;
import php.runtime.lang.BaseWrapper;
import php.runtime.reflection.ClassEntity;

import java.util.Set;

@Reflection.Name(JavaFXExtension.NS + "printing\\UXPrinter")
public class UXPrinter extends BaseWrapper<Printer> {
    interface WrappedInterface {
        @Property String name();
    }

    public UXPrinter(Environment env, Printer wrappedObject) {
        super(env, wrappedObject);
    }

    public UXPrinter(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    private void __construct() {
    }

    @Signature
    public PrinterJob createPrintJob() {
        return PrinterJob.createPrinterJob(getWrappedObject());
    }

    @Signature
    public boolean printWithDialog(Window owner, Node node) {
        PrinterJob printerJob = PrinterJob.createPrinterJob(getWrappedObject());

        if (printerJob.showPrintDialog(owner)) {
            boolean success = printerJob.printPage(node);

            if (success) {
                printerJob.endJob();
            }

            return success;
        } else {
            return false;
        }
    }

    @Signature
    public boolean print(Node node) {
        PrinterJob printerJob = PrinterJob.createPrinterJob(getWrappedObject());
        boolean success = printerJob.printPage(node);

        if (success) {
            printerJob.endJob();
        }

        return success;
    }

    @Signature
    public static Printer getDefault() {
        return Printer.getDefaultPrinter();
    }

    @Signature
    public static Set<Printer> getAll() {
        return Printer.getAllPrinters();
    }

    @Getter
    public PrinterAttributes getAttributes(Environment env, TraceInfo trace) throws Throwable {
        return getWrappedObject().getPrinterAttributes();
    }
}
